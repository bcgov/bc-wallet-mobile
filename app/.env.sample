MEDIATOR_URL=https://f326-207-194-65-204.ngrok.io?c_i=eyJAdHlwZSI6ICJodHRwczovL2RpZGNvbW0ub3JnL2Nvbm5lY3Rpb25zLzEuMC9pbnZpdGF0aW9uIiwgIkBpZCI6ICI2MjQ0ZThiNS0wNWYzLTRhYWItYjM1Yy1lYWVlMWNmZTAyM2MiLCAicmVjaXBpZW50S2V5cyI6IFsiQ3lqM1BHRUJzQ3RyUGFtTTQyRngza3BlYmR2QWdNd1lGejlFS3RmNnlUN3giXSwgImxhYmVsIjogIk1lZGlhdG9yIiwgInNlcnZpY2VFbmRwb2ludCI6ICJodHRwczovL2YzMjYtMjA3LTE5NC02NS0yMDQubmdyb2suaW8ifQ==

# Push notification variables
MEDIATOR_USE_PUSH_NOTIFICATIONS=false
MEDIATOR_LABEL=Mediator