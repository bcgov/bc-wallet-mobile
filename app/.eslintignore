node_modules
__tests__
__mocks__
ios
android

# generated by bob
lib/
bifold-app/
generate-app-package-json.js