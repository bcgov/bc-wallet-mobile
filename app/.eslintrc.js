module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    tsconfigRootDir: __dirname,
    ecmaVersion: 2020, // Allows for the parsing of modern ECMAScript features
    sourceType: 'module', // Allows for the use of imports
  },
  extends: [
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended', // Uses the recommended rules from the @typescript-eslint/eslint-plugin
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    // 'import/ignore': ['react-native'],
    // 'import/extensions': ['.js', '.ts', 'tsx', 'jsx'],
    // 'import/parsers': {
    //   '@typescript-eslint/parser': ['.ts', '.tsx'],
    // },
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {},
    },
  },
  plugins: ['@typescript-eslint', 'import'],
  rules: {
    // Type is enforced by callers. Not entirely, but it's good enough.
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    // Aries protocol defines attributes with snake case.
    '@typescript-eslint/camelcase': 'off',
    '@typescript-eslint/no-use-before-define': ['error', { functions: false, classes: false, variables: true }],
    '@typescript-eslint/explicit-member-accessibility': 'error',
    // Because of early development, we only warn on ts-ignore. In future we want to move to error
    '@typescript-eslint/ban-ts-comment': 'warn',
    'no-console': 'error',
    'import/no-cycle': 'error',
    'import/order': [
      'error',
      {
        groups: ['type', ['builtin', 'external'], 'parent', 'sibling', 'index'],
        alphabetize: {
          order: 'asc',
        },
        'newlines-between': 'always',
      },
    ],
    'import/no-extraneous-dependencies': [
      'error',
      {
        devDependencies: false,
      },
    ],
  },
  globals: {
    require: true,
  },
  overrides: [
    {
      files: ['*.test.*'],
      rules: {
        'no-console': 'off',
        'import/no-extraneous-dependencies': 'off',
      },
    },
    {
      files: ['aries-components/**/*.ts'],
      rules: {
        '@typescript-eslint/no-namespace': 'off',
      },
    },
    {
      files: ['.eslintrc.js', '*.config.js'],
      env: {
        node: true,
      },
    },
  ],
}
