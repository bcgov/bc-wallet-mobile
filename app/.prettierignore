package-lock.json
app.json
ios/
android/
node_modules/
generate-app-package-json.js
src/configs/ledgers/indy/ledgers.json
.gitignore
