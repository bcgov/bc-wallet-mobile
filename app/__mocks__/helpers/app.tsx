import 'reflect-metadata'
import React, { PropsWithChildren, useMemo } from 'react'

import { container } from 'tsyringe'
import { MainContainer, ContainerProvider } from '@hyperledger/aries-bifold-core'

export const BasicAppContext: React.FC<PropsWithChildren> = ({ children }) => {
  const context = useMemo(() => new MainContainer(container.createChildContainer()).init(), [])
  return (
    <ContainerProvider value={context}>{children}</ContainerProvider>
  )
}