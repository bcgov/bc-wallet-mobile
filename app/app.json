{
  "name": "BCWallet",
  "displayName": "BC Wallet"
}
