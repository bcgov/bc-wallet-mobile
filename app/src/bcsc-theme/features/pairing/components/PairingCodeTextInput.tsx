import React, { useState } from 'react'
import { View, StyleSheet, TextInput, TextInputProps } from 'react-native'
import { useTheme } from '@bifold/core'

import { splitSplice } from '@utils/splitSplice'

interface Props extends TextInputProps {
  handleChangeCode: (text: string) => void
}

const PairingCodeTextInput: React.FC<Props> = ({ handleChangeCode, ...textInputProps }) => {
  const [focused, setFocused] = useState(false)
  const { Inputs, maxFontSizeMultiplier, Spacing } = useTheme()
  const [value, setValue] = useState('')

  const styles = StyleSheet.create({
    container: {
      marginVertical: Spacing.sm,
      width: '100%',
    },
  })

  // Capitalize and add a separator midway through the input
  const onChangeText = (text: string) => {
    // Clean and normalize the input (uppercase and remove spaces)
    const normalized = text.toUpperCase().replace(/\s/g, '')

    // Check if user is deleting characters by comparing length with previous value
    const isDeleting = normalized.length < value.replace(/\s/g, '').length

    // Format the display value (add space after first 3 characters for readability)
    // Only format when not deleting or when we still have 4 or more characters
    let formatted = normalized
    if (!isDeleting || normalized.length >= 4) {
      if (normalized.length >= 3) {
        formatted = splitSplice(normalized, 3, 0, ' ')
      }
    }

    // Update the displayed value
    setValue(formatted)

    // Always pass the clean version (without spaces) to the parent
    handleChangeCode(normalized)
  }

  return (
    <View style={styles.container}>
      <TextInput
        maxLength={7}
        autoCapitalize={'characters'}
        autoComplete={'off'}
        autoCorrect={false}
        maxFontSizeMultiplier={maxFontSizeMultiplier}
        style={[Inputs.textInput, focused && Inputs.inputSelected, { textAlign: 'center' }]}
        selectionColor={Inputs.inputSelected.borderColor}
        onFocus={() => setFocused(true)}
        onBlur={() => setFocused(false)}
        value={value}
        onChangeText={onChangeText}
        {...textInputProps}
      />
    </View>
  )
}

export default PairingCodeTextInput
