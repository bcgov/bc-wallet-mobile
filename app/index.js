/* eslint-disable import/no-extraneous-dependencies */
/**
 * @format
 */
import 'react-native-gesture-handler'
// remove these when updated to react-native 0.65.0
import '@formatjs/intl-getcanonicallocales/polyfill'
import '@formatjs/intl-locale/polyfill'
import '@formatjs/intl-pluralrules/polyfill'
import '@formatjs/intl-pluralrules/locale-data/en' // locale-data for en
import '@formatjs/intl-displaynames/polyfill'
import '@formatjs/intl-displaynames/locale-data/en' // locale-data for en
import '@formatjs/intl-listformat/polyfill'
import '@formatjs/intl-listformat/locale-data/en' // locale-data for en
import '@formatjs/intl-numberformat/polyfill'
import '@formatjs/intl-numberformat/locale-data/en' // locale-data for en
import '@formatjs/intl-relativetimeformat/polyfill'
import '@formatjs/intl-relativetimeformat/locale-data/en' // locale-data for en
import '@formatjs/intl-datetimeformat/polyfill'
import '@formatjs/intl-datetimeformat/locale-data/en' // locale-data for en
import '@formatjs/intl-datetimeformat/add-all-tz' // Add ALL tz data

import { NavigationContainer } from '@react-navigation/native'
import React from 'react'
import { AppRegistry } from 'react-native'

import App from './App'
import { name as appName } from './app.json'
import bcwallet from './src/'
const { theme } = bcwallet

const { ColorPallet } = theme

const navigationTheme = {
  dark: true,
  colors: {
    primary: ColorPallet.brand.primary,
    background: ColorPallet.brand.primaryBackground,
    card: ColorPallet.brand.primary,
    text: ColorPallet.grayscale.white,
    border: ColorPallet.grayscale.white,
    notification: ColorPallet.grayscale.white,
  },
}

const Base = () => {
  return (
    <NavigationContainer theme={navigationTheme}>
      <App />
    </NavigationContainer>
  )
}
AppRegistry.registerComponent(appName, () => Base)
