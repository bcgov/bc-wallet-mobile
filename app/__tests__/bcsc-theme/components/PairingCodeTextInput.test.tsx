import { fireEvent, render, waitFor } from '@testing-library/react-native'
import React from 'react'

import { BasicAppContext } from '../../../__mocks__/helpers/app'
import PairingCodeTextInput from '../../../src/bcsc-theme/features/pairing/components/PairingCodeTextInput'

describe('PairingCodeTextInput Component', () => {
  const defaultProps = {
    handleChangeCode: jest.fn(),
    testID: 'pairing-code-input',
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Rendering', () => {
    test('renders correctly', () => {
      const tree = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      waitFor(() => {
        expect(tree).toMatchSnapshot()
      })
    })

    test('renders with correct initial state', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      expect(input.props.value).toBe('')
      expect(input.props.maxLength).toBe(7)
    })
  })

  describe('Input Formatting', () => {
    test('converts lowercase to uppercase', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'abc')

      expect(input.props.value).toBe('ABC')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABC')
    })

    test('adds space after 3 characters', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'ABCD')

      expect(input.props.value).toBe('ABC D')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABCD')
    })

    test('maintains space formatting with full 6-character input', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'ABCDEF')

      expect(input.props.value).toBe('ABC DEF')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABCDEF')
    })

    test('removes existing spaces from input', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'AB C D')

      expect(input.props.value).toBe('ABC D')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABCD')
    })

    test('limits input to 6 characters maximum', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'ABCDEFGHIJ')

      expect(input.props.value).toBe('ABC DEF')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABCDEF')
    })
  })

  describe('Backspace Behavior', () => {
    test('handles backspace when input has 3 characters (before space)', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      fireEvent.changeText(input, 'ABC')
      expect(input.props.value).toBe('ABC')

      fireEvent.changeText(input, 'AB')
      expect(input.props.value).toBe('AB')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('AB')
    })

    test('handles backspace when input has 4 characters (after space)', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      fireEvent.changeText(input, 'ABCD')
      expect(input.props.value).toBe('ABC D')

      fireEvent.changeText(input, 'ABC')
      expect(input.props.value).toBe('ABC')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABC')
    })

    test('handles backspace from full 6-character input', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      fireEvent.changeText(input, 'ABCDEF')
      expect(input.props.value).toBe('ABC DEF')

      fireEvent.changeText(input, 'ABCDE')
      expect(input.props.value).toBe('ABC DE')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABCDE')
    })

    test('handles progressive backspacing through space boundary', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      fireEvent.changeText(input, 'ABCDE')
      expect(input.props.value).toBe('ABC DE')

      fireEvent.changeText(input, 'ABCD')
      expect(input.props.value).toBe('ABC D')

      // backspace to the space
      fireEvent.changeText(input, 'ABC ')
      expect(input.props.value).toBe('ABC')

      fireEvent.changeText(input, 'ABC')
      expect(input.props.value).toBe('ABC')

      fireEvent.changeText(input, 'AB')
      expect(input.props.value).toBe('AB')

      fireEvent.changeText(input, 'A')
      expect(input.props.value).toBe('A')

      fireEvent.changeText(input, '')
      expect(input.props.value).toBe('')
    })
  })

  describe('Edge Cases', () => {
    test('handles empty input', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, '')

      expect(input.props.value).toBe('')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('')
    })

    test('handles input with only spaces', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, '   ')

      expect(input.props.value).toBe('')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('')
    })

    test('handles mixed case input', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'aBcDeF')

      expect(input.props.value).toBe('ABC DEF')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('ABCDEF')
    })

    test('handles input with numbers and letters', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      fireEvent.changeText(input, 'A1B2C3')

      expect(input.props.value).toBe('A1B 2C3')
      expect(defaultProps.handleChangeCode).toHaveBeenCalledWith('A1B2C3')
    })
  })

  describe('Focus States', () => {
    test('handles focus and blur events', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      // Focus and blur states are handled internally, we just ensure no errors occur
      fireEvent(input, 'focus')
      fireEvent(input, 'blur')
    })
  })

  describe('Props Forwarding', () => {
    test('forwards additional TextInput props', () => {
      const additionalProps = {
        placeholder: 'Enter code',
        editable: false,
        accessibilityLabel: 'Pairing code input',
      }

      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} {...additionalProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')
      expect(input.props.placeholder).toBe('Enter code')
      expect(input.props.editable).toBe(false)
      expect(input.props.accessibilityLabel).toBe('Pairing code input')
    })
  })

  describe('Callback Behavior', () => {
    test('calls handleChangeCode with clean values only', () => {
      const { getByTestId } = render(
        <BasicAppContext>
          <PairingCodeTextInput {...defaultProps} />
        </BasicAppContext>
      )

      const input = getByTestId('pairing-code-input')

      fireEvent.changeText(input, 'A')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('A')

      fireEvent.changeText(input, 'AB')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('AB')

      fireEvent.changeText(input, 'ABC')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABC')

      fireEvent.changeText(input, 'ABCD')
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABCD')

      fireEvent.changeText(input, 'ABC DEF') // User somehow enters with space
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABCDEF')

      fireEvent.changeText(input, 'ABCDEFGH') // User pastes too long input
      expect(defaultProps.handleChangeCode).toHaveBeenLastCalledWith('ABCDEF')
    })
  })
})
