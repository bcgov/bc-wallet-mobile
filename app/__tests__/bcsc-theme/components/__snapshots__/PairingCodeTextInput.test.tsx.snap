// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairingCodeTextInput Component Rendering renders correctly 1`] = `
<View
  style={
    {
      "marginVertical": 8,
      "width": "100%",
    }
  }
>
  <TextInput
    autoCapitalize="characters"
    autoComplete="off"
    autoCorrect={false}
    maxFontSizeMultiplier={2}
    maxLength={7}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    selectionColor="#313132"
    style={
      [
        {
          "backgroundColor": "#D3D3D3",
          "borderColor": "#D3D3D3",
          "borderRadius": 4,
          "borderWidth": 1,
          "color": "#313132",
          "fontFamily": "BCSans-Regular",
          "fontSize": 16,
          "padding": 10,
        },
        false,
        {
          "textAlign": "center",
        },
      ]
    }
    testID="pairing-code-input"
    value=""
  />
</View>
`;
